### PR TITLE
Encrypted Value Support

### PR DESCRIPTION
**Pull request by amullan-sage**
Add support to encrypt values prior to generating the query sql to allow for equality, nullity, and in filters on values encrypted in the database.

### CODE REVIEW

File: src/Searchlight/SearchlightEngine.cs [hunk 1]
```diff
@@ -64,6 +65,11 @@ public class SearchlightEngine
         /// DEFAULT: True.
         /// </summary>
         public bool useNoCount { get; set; } = true;
+        
+        /// <summary>
+        /// Encryption implementation to use for encrypted fields.
+        /// </summary>
+        public ISearchlightStringEncryptor Encryptor {get; set; } = null;
```
Comment (ctaylor-sage) [inline src/Searchlight/SearchlightEngine.cs]: nit: formatting, space before `get`

File: src/Searchlight/Parsing/SyntaxParser.cs [hunk 2]
```diff
@@ -410,6 +410,17 @@ private static BaseClause ParseOneClause(SyntaxTree syntax, DataSource source, T
             {
                 return null;
             }
+            
+            if (columnInfo.IsEncrypted && (op != OperationType.Equals && op != OperationType.IsNull && op != OperationType.In))
```
Comment (ctaylor-sage) [inline src/Searchlight/Parsing/SyntaxParser.cs]: feedback: maybe simplify this to a list of allowed ops, so you can do `!EncryptedOperations.Contains(op)`

File: src/Searchlight/Exceptions/InvalidOperation.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    ///
+    /// Example: `(someField gt 5)` where `someField` is an encrypted field.
+    /// </summary>
+    public class InvalidOperation : SearchlightException
+    {
+        public string OriginalFilter { get; internal set; }
+
+        public string FieldName { get; internal set; }
+
+        public string Operation {get; internal set; }
```
Comment (ctaylor-sage) [inline src/Searchlight/Exceptions/InvalidOperation.cs]: nit: formatting

File: src/Searchlight/SearchlightEngine.cs (hunk 1 above)
Comment (amullan-sage) [reply src/Searchlight/SearchlightEngine.cs]: updated

File: src/Searchlight/Parsing/SyntaxParser.cs (hunk 2 above)
Comment (amullan-sage) [reply src/Searchlight/Parsing/SyntaxParser.cs]: updated

File: src/Searchlight/Exceptions/InvalidOperation.cs (hunk 3 above)
Comment (amullan-sage) [reply src/Searchlight/Exceptions/InvalidOperation.cs]: updated

**Review by ctaylor-sage — APPROVED** [review]
Would like some tests but this looks good so gonna approve it

**Review by alexrussak-sage — APPROVED** [review]

**Pull request merged**